File: C# Fundamentals/C# OOP Basics/07. Encapsulation/Ex/P05_PizzaCal/Pizza.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Pizza
{
    private string name;
    private Dough dough;
    private List<Topping> toppings;
    private int toppingCount => this.Toppings.Count;
    private double totalCalories => this.Toppings.Sum(x => x.TotalCalories) + this.Dough.TotalCalories;

    public int ToppingCount
    {
        get { return this.toppingCount; }
    }
    public double TotalCalories
    {
        get
        {
            return this.totalCalories;
        }
    }

    public Pizza(string name)
    {
        this.Name = name;
    }

    public Pizza(string name, Dough dough, List<Topping> toppings)
    {
        this.Name = name;
        this.Dough = dough;
        this.Toppings = new List<Topping>(toppings);
    }

    private string Name
    {
        get
        {
            return this.name;
        }
        set
        {
            if(string.IsNullOrWhiteSpace(value) || value.Length > 15)
            {
                throw new ArgumentException("Pizza name should be between 1 and 15 symbols.");
            }
            this.name = value;
        }
    }
    private Dough Dough
    {
        get
        {
            return this.dough;
        }
        set
        {
            this.dough = value;
        }
    }
    private List<Topping> Toppings
    {
        get
        {
            return this.toppings;
        }
        set
        {
            if(value.Count > 10)
            {
                throw new ArgumentException("Number of toppings should be in range [0..10].");
            }
            this.toppings = value;
        }
    }
}

File: Java/01.Advanced/02.OOP/07.Encapsulation/Ex/src/t_04_pizza_calories/Pizza.java
package t_04_pizza_calories;

import java.util.ArrayList;
import java.util.List;

public class Pizza {
    private String name;
    private Dough dough;
    private final List<Topping> toppings;
    private int numberOfToppings;

    public Pizza(String name, int numberOfToppings) {
        this.setName(name);
        this.setToppings(numberOfToppings);
        toppings = new ArrayList<>();
    }

    public void addTopping(Topping topping) {
        if (toppings.size() >= numberOfToppings) {
            throw new IllegalArgumentException("Number of toppings should be in range [0..10].");
        }
        toppings.add(topping);
    }

    public double getOverallCalories() {
        return this.dough.calculateCalories() + this.toppings.stream().mapToDouble(Topping::calculateCalories).sum();
    }

    public void setDough(Dough dough) {
        this.dough = dough;
    }

    private void setName(String name) {
        if (name == null || name.trim().isEmpty() || name.length() > 15) {
            throw new IllegalArgumentException("t_04_pizza_calories.Pizza name should be between 1 and 15 symbols.");
        }
        this.name = name;
    }

    private void setToppings(int numberOfToppings) {
        if (numberOfToppings < 0 || numberOfToppings > 10) {
            throw new IllegalArgumentException("Number of toppings should be in range [0..10].");
        }
        this.numberOfToppings = numberOfToppings;
    }
}
